### PR TITLE
Add auth-option to authenticate as githubapp

### LINKED ISSUE #112: gh provider - authenticate as app
It would be nice if the operator could avoid using PATs and instead auth as an gh app.
PATs have wide access and are tied to users, and limited API quota compared to apps which are more secure, have narrow permissions, and increased quota.

Here is a controller which wraps this: https://github.com/evryfs/github-actions-runner-operator/blob/master/controllers/githubapi/runnerapi.go#L39
so that one can use app auth or token auth.

Slightly related to #111 (removing username/pw mechanism), while this issue adds another option.

### PR DESCRIPTION
**Pull request by davidkarlsen**
Fixes #112.
Adds support for authenticating as github-app, which offers:
* more narrow-grained permissions (avoiding PATs)
* increased security, using a private key
* increased API-quota

The github-client is now wired up/facilitated through the go-github-app library, which offers caching, metrics etc.

Note: Draft PR - not yet tested, for feedback.

### CODE REVIEW

**Comment by davidkarlsen** [comment]
go-githubapp now has convenience function to configure the transport.
I think this is good to go, but have not yet been able to test it.

**Comment by davidkarlsen** [comment]
That should be it. I've updated the docs and tested it in my 4.6.x cluster, using app-based auth:
```
group-sync-operator-controller-manager-54d5874d76-kh5cq manager 2021-09-06T17:31:04.501Z        INFO    controllers.GroupSync   Beginning Sync  {"groupsync": "group-sync-operator/github-groupsync", "Provider": "github"}
group-sync-operator-controller-manager-54d5874d76-kh5cq manager 2021-09-06T17:31:06.354Z        INFO    controllers.GroupSync   Sync Completed Successfully     {"groupsync": "group-sync-operator/github-groupsync", "Provider": "github", "Groups Created or Updated": 1}
```

CC @sabre1041 @raffaelespazzoli

File: pkg/syncer/syncer.go [hunk 1]
```diff
@@ -15,6 +15,8 @@ const (
 	secretUsernameKey  = "username"
 	secretPasswordKey  = "password"
 	secretTokenKey     = "token"
+	privateKey         = "privateKey"
+	integrationId      = "integrationId"
```
Comment (sabre1041) [inline pkg/syncer/syncer.go:19]: Is there a reason why its called an Integration ID? instead of App ID?

File: pkg/syncer/github.go [hunk 2]
```diff
@@ -145,15 +149,57 @@ func (g *GitHubSyncer) Bind() error {
...
 
-	if tokenSecretFound {
-		ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: string(tokenSecret)})
+	config := githubapp.Config{
+		V3APIURL: *g.Provider.URL,
+		V4APIURL: *g.Provider.URL,
+	}
+
+	opts := []githubapp.ClientOption{
+		githubapp.WithClientUserAgent("redhat-cop/group-sync-operator"),
```
Comment (sabre1041) [inline pkg/syncer/github.go:158]: Can the User Agent be externalized to a constant?

File: README.md [hunk 3]
```diff
@@ -164,11 +167,32 @@ The secret can be created by executing the following command:
 oc create secret generic github-group-sync --from-literal=token=<token>
 ```
 
+##### As a GitHub app
+
+When authenticating as a Github App, the following keys are required:
+* `privateKey` and `integrationId`
+
+
+###### First create a github app
```
Comment (sabre1041) [inline README.md:176]:
```suggestion
###### First create a GitHub App
```

File: README.md [hunk 4]
```diff
@@ -145,14 +145,17 @@ spec:
...
-      credentialsSecret: 
+      credentialsSecret:
         name: github-group-sync
         namespace: group-sync-operator
 ```
 
 #### Authenticating to GitHub
 
-Authentication to GitHub can be performed using an OAuth Personal Access Token. A secret must be created in the same namespace that contains the `GroupSync` resource:
+Authentication to GitHub can be performed using an OAuth Personal Access Token or as an GithubApp, using a secret key and integrationId.
```
Comment (sabre1041) [inline README.md:155]: Change to GitHub App and include links to both OAuth PAT and GitHub App

File: README.md [hunk 5]
```diff
@@ -164,11 +167,32 @@ The secret can be created by executing the following command:
...
+
+When authenticating as a Github App, the following keys are required:
+* `privateKey` and `integrationId`
+
+
+###### First create a github app
+
+In GitHub, go to developer-settings -> github apps. Create a new app, it does not need webhook callbacks.
+Generate a private-key and download it, under "permissions and events" the app will need read-only access to the "Members" permission in the "Organization" section.
+Take note of the "App ID" as you need it for later.
```
Comment (sabre1041) [inline README.md:180]: Notice how "App ID" is used here, but gets added to the secret as an `integrationID`

File: README.md [hunk 6]
```diff
@@ -164,11 +167,32 @@ The secret can be created by executing the following command:
...
 
+##### As a GitHub app
+
+When authenticating as a Github App, the following keys are required:
+* `privateKey` and `integrationId`
+
+
+###### First create a github app
+
+In GitHub, go to developer-settings -> github apps. Create a new app, it does not need webhook callbacks.
```
Comment (sabre1041) [inline README.md:178]: Can these steps be broken out so they are easier to read?

**Review by sabre1041 — CHANGES_REQUESTED** [review]
This is really good. Tested out and successful. Requested changes to documentation and minor things

File: pkg/syncer/syncer.go (hunk 1 above)
Comment (davidkarlsen) [reply pkg/syncer/syncer.go]: For some reason it's called integrationId in the go-api, but let's call it appId which is what is exposed to the user in the gh ui.

**Comment by davidkarlsen** [comment]
@sabre1041 I think all points have been covered now - thanks for the input!

**Review by sabre1041 — APPROVED** [review]
LGTM. Thanks for the contribution. This is a good feature

**Pull request merged**